client.go

```diff
@@ -101,3 +217,346 @@ func (c *Client) SendSTUNRequest(serverIP net.IP, serverPort int) (net.Addr, err
+	)
+	// Trying to authorize.
+	msg, err = stun.Build(
+		stun.TransactionID,
+		stun.NewType(stun.MethodAllocate, stun.ClassRequest),
+		turn.RequestedTransportUDP,
+		&c.username,
+		&c.realm,
+		&nonce,
+		&c.integrity,
+		stun.Fingerprint,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	trRes, err = c.PerformTransaction(msg, c.turnServ, false)
+	if err != nil {
+		return nil, err
+	}
+	res = trRes.Msg
+
+	if res.Type.Class == stun.ClassErrorResponse {
+		var code stun.ErrorCodeAttribute
+		if err = code.GetFrom(res); err == nil {
+			return nil, fmt.Errorf("%s (error %s)", res.Type, code)
+		}
+		return nil, fmt.Errorf("%s", res.Type)
+	}
+
+	// Getting relayed addresses from response.
+	var relayed turn.RelayedAddress
+	if err := relayed.GetFrom(res); err != nil {
+		return nil, err
+	}
+	relayedAddr := &net.UDPAddr{
+		IP:   relayed.IP,
+		Port: relayed.Port,
+	}
+
+	// Getting lifetime from response
+	var lifetime turn.Lifetime
+	if err := lifetime.GetFrom(res); err != nil {
+		return nil, err
+	}
+
+	relayedConn = client.NewUDPConn(&client.UDPConnConfig{
+		Observer:    c,
+		RelayedAddr: relayedAddr,
+		Integrity:   c.integrity,
+		Nonce:       nonce,
+		Lifetime:    lifetime.Duration,
+		Log:         c.log,
+	})
+
+	c.setRelayedUDPConn(relayedConn)
+
+	return relayedConn, nil
+}
+
+// PerformTransaction performs STUN transaction
+func (c *Client) PerformTransaction(msg *stun.Message, to net.Addr, dontWait bool) (client.TransactionResult, error) {
+	trKey := b64.StdEncoding.EncodeToString(msg.TransactionID[:])
+
+	raw := make([]byte, len(msg.Raw))
+	copy(raw, msg.Raw)
+
+	tr := client.NewTransaction(&client.TransactionConfig{
+		Key:      trKey,
+		Raw:      raw,
+		To:       to,
+		Interval: c.rto,
+	})
+
+	c.trMap.Insert(trKey, tr)
+
+	c.log.Tracef("start %s transaction %s to %s", msg.Type, trKey, tr.To.String())
+	_, err := c.conn.WriteTo(tr.Raw, to)
+	if err != nil {
+		return client.TransactionResult{}, err
+	}
+
+	tr.StartRtxTimer(c.onRtxTimeout)
+
+	// If dontWait is true, get the transaction going and return immediately
+	if dontWait {
+		return client.TransactionResult{}, nil
+	}
+
+	res := tr.WaitForResult()
+	if res.Err != nil {
+		return res, res.Err
+	}
+	return res, nil
+}
+
+// OnDeallocated is called when deallocation of relay address has been complete.
+// (Called by UDPConn)
+func (c *Client) OnDeallocated(relayedAddr net.Addr) {
+	c.setRelayedUDPConn(nil)
+}
+
+// HandleInbound handles data received.
+// This method handles incoming packet demultiplex it by the source address
+// and the types of the message.
+// This return a booleen (handled or not) and if there was an error.
+// Caller should check if the packet was handled by this client or not.
+// If not handled, it is assumed that the packet is application data.
+// If an error is returned, the caller should discard the packet regardless.
+func (c *Client) HandleInbound(data []byte, from net.Addr) (bool, error) {
+	var handled bool
```

Comment:
I would return values right away. I think this works great right now, but seems really easy to make a mistake in the future when editing.

```
	case stun.IsMessage(data):
		return true, c.handleSTUNMessage(data, from)
	case len(c.turnServStr) != 0 && from.String() == c.turnServStr:
		if turn.IsChannelData(data) {
			return true, c.handleChannelData(data)
		} else {
		       return true,  fmt.Errorf("unexpected packet from TURN server")
		}
```

Reply:
Agreed. I have modified it as suggested.